testng-core/src/test/java/test/beforegroups/BeforeGroupsTest.java

```diff
@@ -100,6 +103,28 @@ public void ensureBeforeGroupsAreInvokedByDefaultEvenWithoutGrouping() {
     assertThat(AnotherTestClassSample.logs).containsExactlyElementsOf(expectedLogs);
   }
 
+  @Test
+  public void ensureBeforeGroupIsRunBeforeFirstTestInParallelMethodLaunch() {
+    TestNG tng = new TestNG();
+
+    tng.setTestClasses(new Class[] {SampleFor2359.class});
+
+    // bug only exists when running parallel by instances with this flag set to false
+    tng.setParallel(XmlSuite.ParallelMode.METHODS);
+    tng.setThreadCount(3);
+
+    ListenerAdapter adapter = new ListenerAdapter();
+    tng.addListener(adapter);
+
+    tng.run();
+
+    ITestResult beforeGroup = adapter.getPassedConfiguration().iterator().next();
+    adapter
+        .getPassedTests()
+        .forEach(
+            t -> assertThat(t.getStartMillis()).isGreaterThanOrEqualTo(beforeGroup.getEndMillis()));
```

Comment:
I think this assertion should just use `isGreaterThan()` because we are trying to ensure that all the tests started ONLY after the `beforeGroups()` configuration ran to completion.

Reply:
I believe in some circumstances it is possible to get situation when before group stops and test starts absolutes at the same moment

Reply:
@velma - with respect to this test, we are running the beforegroup and then pausing for 2 seconds. So there shouldn't the case that you mention. And if that happens then we have bigger problems because it means that the countdown latch is not being honoured properly right ?

Reply:
Don't see any contradictions. So we are checking that any test starts after or at the same time beforegroup is completed. It should be very rare case (went test's start time equals to beforegroup endtime) but still it is possible (if there are enough CPU time, why not?)

Reply:
@velma - The sample that you have added explicitly sleeps for 2 seconds in the `beforeGroup()` method. So not sure how would a test have the same start time as the configuration method.

Reply:
test does not compare test's start time with configuration method start time, but with **endTime**. So in timeline (BF for before group, and T for tests)
<BF start> -> sleep  -> <BF end>
                                                    <T start>....
                                                    <T start>....
                                                    <T start>....
sleep affect BF end time, but does not affect test start time relative to BF end time

Reply:
![image](https://user-images.githubusercontent.com/896234/163785292-edc96345-9ce2-44fd-9aff-e02d25694e83.png)
can't fight formatting :)